http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/MultiplexedChannelRecord.java

```diff
@@ -31,165 +27,194 @@
-                             long maxConcurrencyPerConnection,
-                             BiConsumer<Channel, MultiplexedChannelRecord> channelReleaser) {
-        this.connectionFuture = connectionFuture;
-        this.childChannels = new ConcurrentHashMap<>(saturatedCast(maxConcurrencyPerConnection));
-        this.availableStreams = new AtomicLong(maxConcurrencyPerConnection);
-        this.channelReleaser = channelReleaser;
+    MultiplexedChannelRecord(Channel connection, long maxConcurrencyPerConnection) {
         this.connection = connection;
+        this.maxConcurrencyPerConnection = maxConcurrencyPerConnection;
+        this.availableChildChannels = new AtomicLong(maxConcurrencyPerConnection);
+    }
+
+    boolean acquireStream(Promise<Channel> promise) {
+        if (reserveStream()) {
+            releaseReservationOnFailure(promise);
+            acquireReservedStream(promise);
+            return true;
+        }
+        return false;
     }
 
-    MultiplexedChannelRecord acquire(Promise<Channel> channelPromise) {
-        availableStreams.decrementAndGet();
-        if (connection != null) {
-            createChildChannel(channelPromise);
-        } else {
-            connectionFuture.addListener((GenericFutureListener<Future<Channel>>) future -> {
-                if (future.isSuccess()) {
-                    connection = future.getNow();
-                    connection.attr(CHANNEL_POOL_RECORD).set(this);
-                    createChildChannel(channelPromise);
-                } else {
-                    channelPromise.setFailure(future.cause());
-                    channelReleaser.accept(connection, this);
+    private void acquireReservedStream(Promise<Channel> promise) {
+        doInEventLoop(connection.eventLoop(), () -> {
+            if (state != RecordState.OPEN) {
+                String message = "Connection received GOAWAY or was closed while acquiring new stream.";
+                promise.setFailure(new IllegalStateException(message));
+                return;
+            }
+
+            Future<Http2StreamChannel> streamFuture = new Http2StreamChannelBootstrap(connection).open();
+            streamFuture.addListener((GenericFutureListener<Future<Http2StreamChannel>>) future -> {
+                warnIfNotInEventLoop(connection.eventLoop());
+
+                if (!future.isSuccess()) {
+                    promise.setFailure(future.cause());
+                    return;
                 }
+
+                Http2StreamChannel channel = future.getNow();
+                childChannels.put(channel.id(), channel);
+                promise.setSuccess(channel);
             });
+        }, promise);
+    }
+
+    private void releaseReservationOnFailure(Promise<Channel> promise) {
+        try {
+            promise.addListener(f -> {
+                if (!promise.isSuccess()) {
+                    releaseReservation();
+                }
+            });
+        } catch (Throwable e) {
```

Comment:
I'm in the habit of doing it, is the main reason. It's unlikely, but it could happen if the listener was null or the promise was null, or we were working with some weird promise. I'd prefer not to rely on the implementation details of the other class to make sure my class functions correctly.